[Woo POS - payments onboarding] Display payments onboarding in POS with existing design

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #14148
<!-- Id number of the GitHub issue this PR addresses. -->

## Why

Now that a feature flag is available for payments onboarding in POS, we want to show the onboarding when it's required in the pre-existing card-present payments flow (onboarding -> card reader connection -> (optional) payment capture).

## How

The onboarding view model is already available in the payment event publisher `cardPresentPaymentService.paymentEventPublisher`, this PR just maps the optional onboarding view model to a Published var `cardPresentPaymentOnboardingViewModel`. I was debating whether to keep it in `TotalsViewModel` or elsewhere like `PointOfSaleDashboardViewModel`, but decided to have it in `TotalsViewModel` to be in the same place as other payment-related variables for now.

With `cardPresentPaymentOnboardingViewModel`, it's used to show a sheet in `PointOfSaleDashboardView` with a navigation bar with a CTA to cancel. I tried showing the onboarding in `posModal` first, but it couldn't be displayed in `NavigationStack` - it will result in the whole POS view being dismissed. I haven't debugged it further, I think it's because `posModal` isn't showing the view as a sheet and there's some issue showing a `NavigationStack` on top of another view. Other views shown with `posModal` aren't using `NavigationStack`.

In order to show the onboarding view as a [sheet's item](https://developer.apple.com/documentation/swiftui/view/sheet(item:ondismiss:content:)), the view model has to conform to `Identifiable`.

From initial testing, I noticed that the onboarding view isn't shown after it has been dismissed once in the app. It was because of this guard statement on the readiness subscription, assuming that there is no other subscription when checking if onboarding is required:

https://github.com/woocommerce/woocommerce-ios/blob/a69a4278c429dfc3162938bb870c26fae5ff5d05/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentOnboardingAdaptor.swift#L34-L36

This works when the merchant proceeds to [complete the onboarding which resets the subscription back to nil](https://github.com/woocommerce/woocommerce-ios/blob/a69a4278c429dfc3162938bb870c26fae5ff5d05/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentOnboardingAdaptor.swift#L62), but it's never reset to nil if the merchant dismisses the onboarding. This prevents them from starting the onboarding check with an optional onboarding display, resulting in a no-op. In https://github.com/woocommerce/woocommerce-ios/pull/14185/commits/dfb07318f557b01e1b7172a82c6a8e6bcc7f8768, I removed this guard statement to allow onboarding checks more than once. Please let me know if I'm missing some reasons for this guard statement.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Stores that haven't completed onboarding

On web:

- Create a new Jurassic Ninja site – include WooCommerce, WooPayments, and WooPayments Dev Tools
- Activate WooPayments plugin
- Link the site with your Jetpack user through WooPayments setup with the WPCOM account in the app, but skip further onboarding setup
- In WooCommerce settings > General, enter a valid US address with a zip code

In the app:

- Launch the app
- Go to Menu > POS Mode
- Tap `Connect your reader` --> after onboarding loading view, the next onboarding step should be shown
- Dismiss the onboarding modal
- Tap `Connect your reader` again --> after onboarding loading view, the same next onboarding step should be shown
- Proceed to complete onboarding, using Test Mode data whenever possible (phone number, 2FA code, upload SSN document) --> after onboarding is complete, tap `Done` on the webview. It should refresh the onboarding state again, and transition to reader connection flow (note: it took me at least 5 minutes to complete this each time 😅 )
- Continue with reader connection --> connection should succeed
- Feel free to add products to cart and then pay for the order

### Stores that have completed onboarding

- Switch to a store that already completed WooPayments onboarding
- Go to Menu > POS Mode
- Tap `Connect your reader` --> reader connection flow should be shown without any onboarding UI

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

I tested on iPad iOS 17.5 simulator, and an iOS 17.5.1 device. Also tested on several self-hosted and JN stores.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/8675163a-5b67-414d-a364-3da3c62d51b8




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

jaclync committed:
TotalsViewModel: add `cardPresentPaymentOnboardingViewModel` published variable for an optional onboarding view model.

jaclync committed:
Implement Hashable/Identifiable necessary for presenting `CardPresentPaymentsOnboardingViewModel` as a SwiftUI sheet item.

jaclync committed:
POS: show onboarding view from view model.

jaclync committed:
Allow multiple calls to `showOnboardingIfRequired` in case the modal is dismissed. `readinessSubscription` is currently set to nil only when completed.

jaclync committed:
Nit: move onboarding view to a function.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/229">20.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Add a test case for `TotalsViewModel.cardPresentPaymentOnboardingViewModel` to ensure that it sets the view model correctly.

jaclync committed:
Nit: remove unnecessary Hashable conformance code as Identifiable is sufficient in `CardPresentPaymentsOnboardingViewModel`.

jaclync committed:
Fix unit test build failure from removing Equatable conformance from `CardPresentPaymentsOnboardingViewModel`.

staskus commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift:
```diff
@@ -31,10 +31,6 @@ final class CardPresentPaymentsOnboardingPresenterAdaptor: CardPresentPaymentsOn
     ///   - completion: Callback when the onboarding is complete
     func showOnboardingIfRequired(from viewController: ViewControllerPresenting,
                                   readyToCollectPayment completion: @escaping () -> Void) {
-        guard readinessSubscription == nil else {
```
Yes, I don't know about this as well 🤔 maybe @joshheald you have a better idea? I assume it was designed to prevent show more than one onboarding at the same time.

@jaclync would setting it `nil` here help prevent issues when onboarding is not completed but we dismiss the onboarding view?
```
    private func showOnboarding(from viewController: ViewControllerPresenting,
                                readyToCollectPayment completion: @escaping () -> Void) {
        let onboardingViewController = CardPresentPaymentsOnboardingViewController(viewModel: onboardingViewModel,
                                                                                   onWillDisappear: { [weak self] in
            self?.readinessSubscription?.cancel() 
           self?.readinessSubscription = nil <----
        })
```

jaclync replied:
> Yes, I don't know about this as well 🤔 maybe @joshheald you have a better idea? I assume it was designed to prevent show more than one onboarding at the same time.
> 
> @jaclync would setting it `nil` here help prevent issues when onboarding is not completed but we dismiss the onboarding view?
> 
> ```
>     private func showOnboarding(from viewController: ViewControllerPresenting,
>                                 readyToCollectPayment completion: @escaping () -> Void) {
>         let onboardingViewController = CardPresentPaymentsOnboardingViewController(viewModel: onboardingViewModel,
>                                                                                    onWillDisappear: { [weak self] in
>             self?.readinessSubscription?.cancel() 
>            self?.readinessSubscription = nil <----
>         })
> ```

Great suggestion on cancelling/nilling the subscription instead -

This seems to be in the [UIKit version](https://github.com/woocommerce/woocommerce-ios/blob/316063cc980ebf2231fa69411940364f3e1706ec/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CardPresentPaymentsOnboardingPresenter.swift#L44-L49) of the onboarding presenter, where it creates the onboarding VC which includes a closure when the view will disappear to cancel the subscription.

In the new adapted version of the onboarding presenter, a view model is emitted instead of a view where we can detect its onDisappear. After a deeper look, I reverted this removal of nil check and use a closure in the `showOnboarding` event to nil out the subscription on view's `onDisappear` in https://github.com/woocommerce/woocommerce-ios/pull/14185/commits/dbdc9a77425b7aab5a4513a9838855e7da9904ab. Lemme know what you think!

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -58,6 +58,9 @@ struct PointOfSaleDashboardView: View {
         .animation(.easeInOut(duration: Constants.connectivityAnimationDuration), value: viewModel.showsConnectivityError)
         .background(Color.posPrimaryBackground)
         .navigationBarBackButtonHidden(true)
+        .sheet(item: $totalsViewModel.cardPresentPaymentOnboardingViewModel) { viewModel in
```
From a usability point of view, I dismissed the Stripe onboarding web view at least a few times by accidentally tapping on the background. This forced me to re-enter Stripe onboarding details. I don't know if there's an easy way to prevent it on Swift UI.

jaclync replied:
Good point, it could be a frustrating experience having filled out a bunch of info in the Stripe form and then the form is accidentally closed from dragging or tapping outside of the modal. In https://github.com/woocommerce/woocommerce-ios/pull/14185/commits/f754d0a7ecb6a9f5fbbd2f19fc7e09b64eb63fc2, I added `interactiveDismissDisabled` to the sheet so that it can't be dismissed by these built-in gestures.

staskus replied:
Thanks for the fix!

staskus reviewed (commented):
> Stores that haven't completed onboarding

I tried to do a full onboarding on the web and on the app a few times and eventually I got stuck on this screen. I seem not to pass some verification, although WooCommerce settings display onboarding as completed. I know it's unrelated to these changes but I wasn't able to verify it for now, I'll need to understand what's wrong for me 😑

<img width="1180" alt="image" src="https://github.com/user-attachments/assets/3a5b3f99-d369-41ad-a5b5-a851f7b39279">

https://github.com/user-attachments/assets/ed59f72e-7d8a-472f-ad7e-5859265ae0c1

> Stores that have completed onboarding

✅

jaclync commented:
> I know it's unrelated to these changes but I wasn't able to verify it for now, I'll need to understand what's wrong for me 😑

Can you try checking the UI you see the same onboarding error from Menu > Payments? Just to double-check the onboarding state is consistent in/out of POS. I got into this state after deactivating the WooPayments plugin as well. I'll look into why the error message isn't very helpful here tomorrow, and try reproducing your case.

staskus commented:
@jaclync I retried doing onboarding but inputting the exact same data you did and it all worked well. I'm not sure why the data I used to input before is not working anymore. Either way, I was able to onboard on POS successfully 🎉

jaclync committed:
Revert the removal of `readinessSubscription` nil check and set the subscription to nil from a callback in the `showOnboarding` event instead.

jaclync committed:
Merge branch 'trunk' into feat/14148-pos-payments-onboarding

* trunk: (129 commits)
  Update RELEASE-NOTES.txt
  Localization update
  Add unit tests for handleTap method
  Add unit tests for WooShippingServiceCardViewModel
  Fix extraInfoLabel capitalization and when there is no extra info
  Fix insurance string when coverage is 0
  Show weight field if saving template and update tests
  Update release notes
  Update indentation
  Extract unitInputView method
  Add weight field to custom package creation
  Update TopTabView.swift
  Update RELEASE-NOTES.txt
  clear all items in memory on reload
  Remove newitems count check
  Assure unique productIDs. Test current page behavior
  Clean up for review
  ammend ab1e3db7 make compile
  Automate next page load on scroll
  nextPage initial logic. fetch next on tap
  ...

jaclync committed:
Remove code for future task that breaks the build.

jaclync committed:
Disable built-in gestures that can dismiss a sheet so that the onboarding sheet cannot be dismissed accidentally.

jaclync committed:
Fix unit test build failure.

jaclync commented:
@staskus I just updated the PR for another pass!

> I retried doing onboarding but inputting the exact same data you did and it all worked well. I'm not sure why the data I used to input before is not working anymore. Either way, I was able to onboard on POS successfully 🎉

Nice! I didn't see the form info in your recording to reproduce, but it's odd that the wp-admin Payments tab doesn't show any error either. Sometimes some missing fields like tax ID could result in an onboarding error, and there might be some additional requirements depending on the store type like non-profit.

staskus reviewed (approved):
🎉 All works well!

Thanks for addressing the cancellation.

joshheald commented:
@jaclync I've not reviewed the code, but from your notes:

> With cardPresentPaymentOnboardingViewModel, it's used to show a sheet in PointOfSaleDashboardView with a navigation bar with a CTA to cancel. I tried showing the onboarding in posModal first, but it couldn't be displayed in NavigationStack - it will result in the whole POS view being dismissed. I haven't debugged it further, I think it's because posModal isn't showing the view as a sheet and there's some issue showing a NavigationStack on top of another view. Other views shown with posModal aren't using NavigationStack.

Yes, when I added the onboarding support we were still using a sheet to show all the payments screens, so this will have worked at the time. When we moved over to POSModals we needed to change the dismissal code in the adaptor. It's probably the same thing that's needed here.

Dismissal of POSModals for payment alerts is [provided to the view model by `PointOfSaleCardPresentPaymentEventPresentationStyle`](https://github.com/woocommerce/woocommerce-ios/blob/915c1a25bf9d737500bf5d59e99968b0c8a8f14c/WooCommerce/Classes/POS/Presentation/Card%20Present%20Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift#L35), which comes from the [`TotalsViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/915c1a25bf9d737500bf5d59e99968b0c8a8f14c/WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift#L393), as that's what presented the modal in the first place.

To avoid the system dismiss call, we had to change the existing code to use generics, so that we could pass it a new `CardPresentPaymentsAlertPresenterAdaptor` which uses `CardPresentPaymentEventDetails` (via generics) instead of the legacy [`CardPresentPaymentsModalViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift), the implementations of which have the calls to dismiss.

I don't think the legacy onboarding view models have the same calls to a system dismiss, but if needed we could use a similar approach here and inject the dismiss function from wherever the posModal is presented.

jaclync commented:
> Yes, when I added the onboarding support we were still using a sheet to show all the payments screens, so this will have worked at the time. When we moved over to POSModals we needed to change the dismissal code in the adaptor. It's probably the same thing that's needed here.
> 
> Dismissal of POSModals for payment alerts is [provided to the view model by `PointOfSaleCardPresentPaymentEventPresentationStyle`](https://github.com/woocommerce/woocommerce-ios/blob/915c1a25bf9d737500bf5d59e99968b0c8a8f14c/WooCommerce/Classes/POS/Presentation/Card%20Present%20Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift#L35), which comes from the [`TotalsViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/915c1a25bf9d737500bf5d59e99968b0c8a8f14c/WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift#L393), as that's what presented the modal in the first place.
> 
> To avoid the system dismiss call, we had to change the existing code to use generics, so that we could pass it a new `CardPresentPaymentsAlertPresenterAdaptor` which uses `CardPresentPaymentEventDetails` (via generics) instead of the legacy [`CardPresentPaymentsModalViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift), the implementations of which have the calls to dismiss.
> 
> I don't think the legacy onboarding view models have the same calls to a system dismiss, but if needed we could use a similar approach here and inject the dismiss function from wherever the posModal is presented.

Thanks for the helpful info @joshheald, I'm leaving the change to use `posModal` as a future task https://github.com/woocommerce/woocommerce-ios/issues/14209 and I'll look into more next week. Just one clarification, can we use `NavigationStack` in `posModal`, or the navigation bar UI should be implemented manually (like a view with bar height with some title and CTA inside)? From my testing, having a `NavigationStack` in the POS modal pops the whole POS view when the `posModal` is triggered to be shown.

jaclync commented:
Merging the PR now to make future work easier! The onboarding feature is behind a feature flag.

joshheald commented:
> Thanks for the helpful info @joshheald, I'm leaving the change to use posModal as a future task https://github.com/woocommerce/woocommerce-ios/issues/14209 and I'll look into more next week. Just one clarification, can we use NavigationStack in posModal, or the navigation bar UI should be implemented manually (like a view with bar height with some title and CTA inside)? From my testing, having a NavigationStack in the POS modal pops the whole POS view when the posModal is triggered to be shown.

I've never tried... but I wouldn't invest time in that, personally. Navigation bars will look out of place in our POS designs, and we don't need them for onboarding (e.g. when we show onboarding inline in the IPP payment flow, rather than manage card readers, we don't show a nav bar.)

I can't remember if a navigation stack is required though, it might be for some of the actions. I still think that any time spent making it work is better spent on moving towards the newer style (even though we don't have designs for it), than supporting the older one.
